tiledb/sm/storage_manager/consolidator.cc

```diff
@@ -374,6 +624,104 @@ void Consolidator::free_buffers(
   delete[] buffer_sizes;
 }
 
+template <class T>
+Status Consolidator::compute_next_to_consolidate(
+    const ArraySchema* array_schema,
+    const std::vector<FragmentInfo>& fragments,
+    std::vector<FragmentInfo>* to_consolidate,
+    T* union_non_empty_domains,
+    bool* coincides) const {
+  // Preparation
+  to_consolidate->clear();
+  auto min = config_.min_frags_;
+  min = (uint32_t)((min > fragments.size()) ? fragments.size() : min);
+  auto max = config_.max_frags_;
+  max = (uint32_t)((max > fragments.size()) ? fragments.size() : max);
+  auto size_ratio = config_.size_ratio_;
+
+  // Trivial case - no fragments
+  if (max == 0)
+    return Status::Ok();
+
+  // Prepare the dynamic-programming matrix. The rows are from 1 to max
+  // and the columns represent the fragments in `fragments`.
+  std::vector<std::vector<uint64_t>> m;
+  auto col_num = fragments.size();
+  auto row_num = max;
+  m.resize(row_num);
+  for (auto& row : m)
+    row.resize(col_num);
+
+  // Entry m[i][j] contains the collective size of fragments
+  // fragments[j], ..., fragments[j+i]. If the size ratio
+  // of any adjacent pair in the above list is smaller than the
+  // defined one, then the size sum of that entry is infinity (UINT64_MAX).
+  for (size_t i = 0; i < row_num; ++i) {
+    for (size_t j = 0; j < col_num; ++j) {
+      if (i == 0) {  // In the first row we store the sizes of `fragments`
+        m[i][j] = fragments[j].fragment_size_;
+      } else if (i + j >= col_num) {  // Non-valid entries
+        m[i][j] = UINT64_MAX;
+      } else {  // Every other row is computed using the previous row
+        auto ratio = (float)fragments[i + (j - 1)].fragment_size_ /
+                     fragments[i + j].fragment_size_;
+        ratio = (ratio <= 1.0f) ? ratio : 1.0f / ratio;
+        if (ratio >= size_ratio)
+          m[i][j] = m[i - 1][j] + fragments[i + j].fragment_size_;
+        else
+          m[i][j] = UINT64_MAX;
+      }
+    }
+  }
+
+  // Choose the maximal set of fragments with cardinality in [min, max]
+  // with the minimum size
+  uint64_t min_size = UINT64_MAX;
+  size_t min_col = 0;
+  for (int i = row_num - 1; (i >= 0 && i >= (int)min - 1); --i) {
+    min_size = UINT64_MAX;
+    for (size_t j = 0; j < col_num; ++j) {
+      // Update the min size if the new size is more than 25% smaller.
+      // This is to give preference to earlier fragment sets, in case
+      // the user writes in *approximately* equal batches. Otherwise,
+      // fragment sets in the middle of the timeline may get consolidated,
+      // which will hinder the next step of consolidation (which will
+      // select some small and some big fragments).
+      if (min_size == UINT64_MAX || m[i][j] < (min_size / 1.25)) {
```

Comment:
Great catch - I was sure you were going to see that. :)

This is to prevent a subtle but important case, which I encountered when I was performing some perf analysis locally. Suppose that you have 10 fragments, each of **roughly** equal size. I am saying roughly, because even if you write the exact same number of cells in each fragment, if you use compression the actual fragment sizes may be slightly different. 

Suppose also that we consolidate 5 at a time. if the algorithm chooses to consolidate `2-6` first, then the algorithm will next choose `1, 2-6, 7, 8, 9` into `1-9` (or `2-6, 7, 8, 9,10` into `2-10`), and then `1-9, 10`. Observe the imbalance. The ideal scenario would have been `1-5`, `6-10` and then `1-5, 6-10`. This condition biases towards "alignment", rather than "old" fragments.